Added Touch Behavior Migration Guide (#399)
MicrosoftDocs/CommunityToolkit: docs/maui/behaviors/touch-behavior.md

## API

You can find the source code for `TouchBehavior` over on the [.NET MAUI Community Toolkit GitHub repository](https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui/Behaviors/TouchBehavior.shared.cs).

## Migrating From Xamarin Community Toolkit

> In the [Xamarin Community Toolkit](https://github.com/xamarin/XamarinCommunityToolkit), there was the [`TouchEffect`](https://learn.microsoft.com/en-us/dotnet/api/xamarin.communitytoolkit.effects.toucheffect?view=xamarin-community-toolkit-sdk) if you are upgrading an app from Forms to Maui there are a number of changes that you need to be aware of.

Reviewer: As we are on the Microsoft docs page we need to use the full names

```suggestion
In the [Xamarin Community Toolkit](https://github.com/xamarin/XamarinCommunityToolkit), there was the [`TouchEffect`](https://learn.microsoft.com/en-us/dotnet/api/xamarin.communitytoolkit.effects.toucheffect?view=xamarin-community-toolkit-sdk) if you are upgrading an app from Xamarin.Forms to .NET MAUI there are a number of changes that you need to be aware of.
```

Author: Fully qualified all names

> The XCT implementation was an attached effect and was applied to an element in the following way:

Reviewer: ```suggestion
The Xamarin Community Toolkit implementation was an attached effect and was applied to an element in the following way:
```

```xaml
<StackLayout Orientation="Horizontal"
            HorizontalOptions="CenterAndExpand"
            xct:TouchEffect.AnimationDuration="250"
            xct:TouchEffect.AnimationEasing="{x:Static Easing.CubicInOut}"
            xct:TouchEffect.PressedScale="0.8"
            xct:TouchEffect.PressedOpacity="0.6"
            xct:TouchEffect.Command="{Binding Command}">
    <BoxView Color="Gold" />
    <Label Text="The entire layout receives touches" />
    <BoxView Color="Gold"/>
</StackLayout>
```

You should be aware of the 2 following major changes to this effect:
1. `TouchBehavior` is now a `PlatformBehavior`, not an `AttachedEffect`

> 2. The `BindingContext is not automatically set

Reviewer: ```suggestion
2. The `BindingContext` is not automatically set
```

Reviewer 2: Can we maybe also write a little bit more about what that means exactly? More experienced developers will know, newcomers will maybe have trouble understanding what the symptoms are they might be seeing and what problem this solves.

Doesn't have to be super extensive, just something like: that means data-binding will not automatically work and pick up values in your view models.

I see you're expanding on this a little bit later? Always good to add a reference.

Author: I've turned this article into 2 sections and put links to each in this bullet list so it is clear about each point

### Migrating To Platform Behavior

The behavior is now applied to the elements behavior collection, see [Platform Behaviors](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/behaviors?view=net-maui-8.0#platform-behaviors) for more information.

```xaml
<HorizontalStackLayout HorizontalOptions="CenterAndExpand" VerticalOptions="Center">
    <HorizontalStackLayout.Behaviors>
        <mct:TouchBehavior
            DefaultAnimationDuration="250"
            DefaultAnimationEasing="{x:Static Easing.CubicInOut}"
            PressedOpacity="0.6"
            PressedScale="0.8" />
    </HorizontalStackLayout.Behaviors>

    <ContentView
        BackgroundColor="Gold"
        HeightRequest="100"
        WidthRequest="10" />
    <Label
        LineBreakMode="TailTruncation"
        Text="The entire layout receives touches"
        VerticalOptions="Center" />
    <ContentView
        BackgroundColor="Gold"
        HeightRequest="100"
        WidthRequest="10" />
</HorizontalStackLayout>
```

### Setting The TouchBehavior Binding Context

> As noted in the previous section, the `BindingContext` of the behavior is not set automatically, the reason for this is outlined in the Maui documentation for [Behaviors](https://learn.microsoft.com/en-gb/dotnet/maui/fundamentals/behaviors?view=net-maui-8.0#create-a-net-maui-behavior):

Reviewer: ```suggestion
As noted in the previous section, the `BindingContext` of the behavior is not set automatically, the reason for this is outlined in the .NET MAUI documentation for [Behaviors](https://learn.microsoft.com/en-gb/dotnet/maui/fundamentals/behaviors?view=net-maui-8.0#create-a-net-maui-behavior):
```

| NormalOpacity                     | DefaultOpacity                                       |
| NormalTranslationX                | DefaultTranslationX                                  |
| NormalTranslationY                | DefaultTranslationY                                  |
| NormalRotation                    | DefaultRotation                                      |
| NormalRotationX                   | DefaultRotationX                                     |
| NormalRotationY                   | DefaultRotationY                                     |
| NormalAnimationDuration           | DefaultAnimationDuration                             |
| NormalAnimationEasing             | DefaultAnimationEasing                               |
| NormalBackgroundImageSource       | DefaultImageSource (Moved to `ImageTouchBehavior`)   |
| NormalBackgroundImageAspect       | DefaultImageAspect (Moved to  `ImageTouchBehavior` ) |

### TouchBehavior is now implemented as a `PlatformBehavior`

In the Xamarin Community Toolkit the `TouchEffect` was implemented as an `AttachedEffect`. To use the effect you would use the attached properties and apply to any `VisualElement`

> In .NET Maui the `TouchBehavior` is implemented as a `PlatformBehavior` which is now applied to the elements behavior collection, see [Platform Behaviors](/dotnet/maui/fundamentals/behaviors#platform-behaviors) for more information.

Author: Good spot, I've updated every instance of `Maui` to `MAUI`

> > Note: By default in .NET Maui `PlatformBehavior`s will not set a BindingContext, this is because Behaviors can be shared in styles. The `TouchBehavior` will set it's `BindingContext` equal to the `VisualElement` it is applied to. This means that you should not share the `TouchBehavior` between elements via styles.

Author: Good spot, updated

## API

You can find the source code for `TouchBehavior` over on the [.NET MAUI Community Toolkit GitHub repository](https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui/Behaviors/TouchBehavior.shared.cs).

> ## Migrating From Xamarin Community Toolkit

Reviewer 2: Not sure if it's relevant but should you include a full example?

This is how your Effect looked like in Xamarin, this is the same Behavior in .NET MAUI so you can really easily spot the differences? Or are they that small that this doesn't make sense?

Author: I'm happy to add an extra example, in my mind I put the xct example at the start of the example and then I have the equivalent maui example later. I'll put them together so its clearer for the reader :)

Author: I've merged the XCT & Maui examples into the same section so its clear that A migrates to B

> In the [Xamarin Community Toolkit](https://github.com/xamarin/XamarinCommunityToolkit), there was the [`TouchEffect`](https://learn.microsoft.com/en-us/dotnet/api/xamarin.communitytoolkit.effects.toucheffect?view=xamarin-community-toolkit-sdk) if you are upgrading an app from Forms to Maui there are a number of changes that you need to be aware of.

Reviewer 2: ```suggestion
In the [Xamarin Community Toolkit](https://github.com/xamarin/XamarinCommunityToolkit), there was the [`TouchEffect`](/dotnet/api/xamarin.communitytoolkit.effects.toucheffect?view=xamarin-community-toolkit-sdk) if you are upgrading an app from Forms to Maui there are a number of changes that you need to be aware of.
```

I think this should work and actually we'll get a warning otherwise to not do hard links.

As an aside: it's always good at the very least to remove the `en-us` identifier so the reader will be redirected to their own localized version if there is one.

The XCT implementation was an attached effect and was applied to an element in the following way:

```xaml
<StackLayout Orientation="Horizontal"
            HorizontalOptions="CenterAndExpand"
            xct:TouchEffect.AnimationDuration="250"
            xct:TouchEffect.AnimationEasing="{x:Static Easing.CubicInOut}"
            xct:TouchEffect.PressedScale="0.8"
            xct:TouchEffect.PressedOpacity="0.6"
            xct:TouchEffect.Command="{Binding Command}">
    <BoxView Color="Gold" />
    <Label Text="The entire layout receives touches" />
    <BoxView Color="Gold"/>
</StackLayout>
```

You should be aware of the 2 following major changes to this effect:
1. `TouchBehavior` is now a `PlatformBehavior`, not an `AttachedEffect`
2. The `BindingContext is not automatically set

### Migrating To Platform Behavior

> The behavior is now applied to the elements behavior collection, see [Platform Behaviors](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/behaviors?view=net-maui-8.0#platform-behaviors) for more information.

Reviewer 2: ```suggestion
The behavior is now applied to the elements behavior collection, see [Platform Behaviors](/dotnet/maui/fundamentals/behaviors#platform-behaviors) for more information.
```

```xaml
<HorizontalStackLayout HorizontalOptions="CenterAndExpand" VerticalOptions="Center">
    <HorizontalStackLayout.Behaviors>
        <mct:TouchBehavior
            DefaultAnimationDuration="250"
            DefaultAnimationEasing="{x:Static Easing.CubicInOut}"
            PressedOpacity="0.6"
            PressedScale="0.8" />
    </HorizontalStackLayout.Behaviors>

    <ContentView
        BackgroundColor="Gold"
        HeightRequest="100"
        WidthRequest="10" />
    <Label
        LineBreakMode="TailTruncation"
        Text="The entire layout receives touches"
        VerticalOptions="Center" />
    <ContentView
        BackgroundColor="Gold"
        HeightRequest="100"
        WidthRequest="10" />
</HorizontalStackLayout>
```

### Setting The TouchBehavior Binding Context

> As noted in the previous section, the `BindingContext` of the behavior is not set automatically, the reason for this is outlined in the Maui documentation for [Behaviors](https://learn.microsoft.com/en-gb/dotnet/maui/fundamentals/behaviors?view=net-maui-8.0#create-a-net-maui-behavior):

Reviewer 2: ```suggestion
As noted in the previous section, the `BindingContext` of the behavior is not set automatically, the reason for this is outlined in the Maui documentation for [Behaviors](/dotnet/maui/fundamentals/behaviors#create-a-net-maui-behavior):
```

> > .NET MAUI does not set the BindingContext of a behavior, because behaviors can be shared and applied to multiple controls through styles.

Reviewer 2: ```suggestion
> .NET MAUI does not set the `BindingContext` of a behavior, because behaviors can be shared and applied to multiple controls through styles.
```

Author: Done
